File: python/max_pairwise_product.py
# Reimplementing the max_pairwise_product.cpp in Python
# See the repo for cpp
import random # needed for stress test

def max_pairwise_product(num_list):
  """
  The simplest possible solution.
  The function simply multiplies all the possible pairs,
  and chooses the largest result.
  
  Very slow.
  """
  result = 0
  n = len(num_list)
  for i in range(n):
    for j in range(i+1, n):
      if (num_list[i] * num_list[j]) > result:
        result = num_list[i] * num_list[j]
  return result
  

# The function above rewritten for execution efficiency
def max_pairwise_product_fast(num_list):
  """
  Iteraates through the list twice to find indeces of two largest numbers.
  Then nultiplies the numbers and returns the product.
  """
  n = len(num_list)
  
  max_index1 = -1
  for i in range(n):
    if (max_index1 == -1) or (num_list[i] > num_list[max_index1]):
      max_index1 = i
  
  max_index2 = -1
  for j in range(n):
    if (j != max_index1) and ((max_index2 == -1) or (num_list[j] > num_list[max_index2])):
      max_index2 = j
  
  return num_list[max_index1] * num_list[max_index2]

# IMPLEMENT STRESS TEST

#while True:
#  len_list = 1000
#  random_list = random.sample(range(0, 10000), len_list)
#  print(random_list)
#
#  res1 = max_pairwise_product(random_list)
#  res2 = max_pairwise_product_fast(random_list)
#
#  if res1 != res2:
#    print(f"Wrong answer {res1} {res2}")
#  else:
#    print("OK")
  
  
  
  
n = int(input()) # actually we don't need this here

num_list_as_str = input()
#print(num_list_as_str)
num_list_as_list_of_str = num_list_as_str.split()
#print(num_list_as_list_of_str)
num_list = [int(i) for i in num_list_as_list_of_str]

result = max_pairwise_product_fast(num_list)
print(result)

    
      

File: course_1/week1/cpp/max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <cstdlib> //used to generate random functions

using std::vector;
using std::cin;
using std::cout;


//NAIVE IMPLEMENTATION - WORKS CORRECT BUT TOO SLOW

long long max_pairwise_product(const vector<int>& numbers){
  long long result = 0;
  int n = numbers.size();
  for (int i = 0; i < n; i++){
    for (int j = i + 1; j < n; j++){
      if (((long long)numbers[i]) * numbers[j] > result){
        result = ((long long)numbers[i]) * numbers[j];
      }
    }
  }
  return result;
}

//IMPLEMENTATION USED FOR SUBMISSION
long long max_pairwise_product_fast(const vector<int>& numbers){
  int n = numbers.size();
  
  int max_index1 = -1;
  for (int i = 0; i < n; i++){
    if ((max_index1 == -1) || (numbers[i] > numbers[max_index1]))
      max_index1 = i;
  }
  
  int max_index2 = -1;
  for (int j = 0; j < n; j++){
    if ((j != max_index1) && ((max_index2 == -1) || (numbers[j] > numbers[max_index2])))
      max_index2 = j;
    
    //cout << max_index1 << ' ' << max_index2 << "\n";
  }
  return ((long long)(numbers[max_index1])) * numbers[max_index2];
}

int main(){
  
  //STRESS TEST
  /*
  while (true){
    int n = rand() % 1000 + 2;
    cout << n << "\n";
    vector<int> a;
    for (int i = 0; i < n; i++){
      a.push_back(rand() % 100000);
    }
    for (int i = 0; i < n; i++){
      cout << a[i] << ' ';
    }
    cout << "\n";
    long long res1 = max_pairwise_product(a);
    long long res2 = max_pairwise_product_fast(a);
    if (res1 != res2){
      cout << "Wrong answer: " << res1 << ' ' << res2 << "\n";
      break;
    }
    else {
      cout << "OK\n";
    }
  }
  */
  //END OF STRESS TEST
  
  int n;
  cin >> n;
  vector<int> numbers(n);
  for (int i = 0; i < n; i++){
    cin >> numbers[i];
  }
  
  long long result = max_pairwise_product_fast(numbers);
  cout << result << "\n";
  
  
  return 0;
}
